prolog/scasp/dyncall.pl

```diff
@@ -511,19 +535,21 @@
 %   Declare Spec, a comma list  of   _heads_  to be _abducible_, meaning
 %   they can both be in or outside the model.
 
-abducible(M:(A,B)) =>
-    abducible(M:A),
-    abducible(M:B).
-abducible(M:Head), callable(Head) =>
+abducible(Spec) :- abducible(Spec, false).
+
+abducible(M:(A,B), Pos) =>
+    abducible(M:A, Pos),
+    abducible(M:B, Pos).
+abducible(M:Head, Pos), callable(Head) =>
     abducible_rules(Head, Rules),
-    @(maplist(assertz, Rules), M).
+    maplist(scasp_assert_into(M, Pos), Rules),
+    @(assertz((:- discontiguous(('abducible$'/1, 'abducible$$'/1)))), M).
```

Comment:
>  You cannot assert a directive; this will simply assert a rule for (:-)/1 (which should probably be forbidden).

Thanks, that makes sense. Note that as far as I understand the issue you describe was in effect before this PR, we had:

```prolog
abducible(M:Head), callable(Head) =>
    abducible_rules(Head, Rules),
    @(maplist(assertz, Rules), M).

abducible_rules(HeadIn,
                [ (Head                 :- not AHead, 'abducible$'(Head)),
                  (AHead                :- not Head),
                  ('abducible$'(Head)   :- not 'abducible$$'(Head)),
                  ('abducible$$'(Head)  :- not 'abducible$'(Head)),
                  (:- discontiguous(('abducible$'/1, 'abducible$$'/1)))
                ]) :-
    ab_heads(HeadIn, Head, AHead).
```
So the term `:- discontiguous(('abducible$'/1, 'abducible$$'/1)` was passed as the argument of `assertz/1` it seems.

Because this PR added source positions for dynamic clauses, which should apply also to the clauses that make up an abducible predicate, I wanted to replace the plain `assertz/1`s with `scasp_assert/2` so `Pos` is asserted as the source position of the created clause. But `scasp_assert/2` (which `scasp_assert_into/3` delegates to) doesn't handle directives well, because it treats them as s(CASP) global constraints rather than Prolog directives, so I moved the `discontiguous` directive out of the `maplist` argument list in order to preserve the preexisting situation in which it I figured it was passed to `assertz/1`.

Also note that current state with your latest fix may also not be correct, as because now we pass the `:- discontiguous/1` term to `scasp_assert_into/3` and as I mentioned above this results in the directive being treated as a global constraint.

Maybe we should just call `@(discontiguous(('abducible$'/1, 'abducible$$'/1)), M)` directly instead of asserting it? WDYT?

Reply:
You are probably right.   The simple call is

    discontiguous(M:(...)).

No use for the complicated @/2 rules here.  Pushed 217a7185823dda891b08eac6b9c5ff316bc4b31c